Revamp notice appearing in AMP Enabled toggle, revamp how plugin determines whether AMP can be disabled

westonruter opened the pull request:
## Summary

Fixes #1864.
Fixes #2724.
FIxes #4593.

This will require heavy rewriting of https://amp-wp.org/documentation/how-the-plugin-works/amp-plugin-serving-strategies/

### AMP Unavailable Notice

* When a post cannot be served as AMP, show the notice as informational rather than as a warning.
* Start the notice with **AMP Unavailable** to give context for what the errors are.
* When a link appears in a notice, have it open in a new window.
* In the Classic editor, the notice continues to be displayed only when clicking the "Edit" link. In Gutenberg, the notice is displayed by default but inside of a `details` element which is expanded to show the notices.
* If the user cannot `manage_options` then _not_ AMP toggle is displayed and no notice is shown. Non-administrators will not see the toggle nor any notice if AMP is unavailable, sine they do not have the capability to take any action anyway.

Change | Gutenberg | Classic 
-------|------------|--------
Before | ![image](https://user-images.githubusercontent.com/134745/87219246-66ceaa80-c30e-11ea-82f3-661cf8ca3041.png) | (After clicking the edit link) ![image](https://user-images.githubusercontent.com/134745/87219311-e2c8f280-c30e-11ea-88c1-5bd1f2b79860.png)
After (Collapsed) | ![image](https://user-images.githubusercontent.com/134745/87219745-cfb82180-c312-11ea-923a-e0e708dc0a48.png) | (Looks same as before) ![image](https://user-images.githubusercontent.com/134745/87219756-eb232c80-c312-11ea-934b-17c855003e7c.png)
After (Expanded)    | ![image](https://user-images.githubusercontent.com/134745/87219356-17d54500-c30f-11ea-91bf-f1f387cd6907.png) | ![image](https://user-images.githubusercontent.com/134745/87219341-012eee00-c30f-11ea-93ca-81e7dbdb3a91.png)
After (Non-admin) | ![image](https://user-images.githubusercontent.com/134745/87219807-4a813c80-c313-11ea-8d1b-e7f967bf5652.png) | ![image](https://user-images.githubusercontent.com/134745/87219799-389f9980-c313-11ea-9a23-26d8032982ad.png)

### DB Upgrade

* Wait until the end of `init` to perform the DB upgrade. This ensures that themes have declared theme support and post type support which we'll then persist in the DB going forward.

### Theme Support

* When `add_theme_support('amp')` is present, default to `paired` as being `true` as opposed to `false`. This only impacts the default `theme_support` value the first time the plugin is activated.
* The `templates_supported` flag now only affects the default value for the options for `all_templates_supported` and `supported_templates` (more below).

### Post Types

* Eliminate the use of `amp` post type support to programmatically _override_ the ability for the user to enable or disable a AMP for a given post type.
* ~Migrate to storing options as associative arrays: store `supported_post_types` option as associative array mapping the post type to a boolean. Migrate previous representation of string array.~
* The first time the plugin is activated, when the `amp-options` option is empty: use the `amp` post type support to populate the default `supported_post_types`.

### Templates

* Eliminate the use of `amp` theme support to programmatically _override_ the ability for the user to control whether all templates are supported and which specific templates are supported.
* Eliminate the ability to use the `amp_supportable_templates` filter to programmatically override whether a template supported in AMP. If a filter tries to programmatically supply `supported` or `immutable` then `_doing_it_wrong()` is raised.
* ~Migrate to storing options as associative arrays: store `supported_templates` option as an associative array mapping template ID to a boolean. Migrate previous representation of string array.~
* The first time the plugin is activated, when the `amp-options` option is empty: Use the `templates_supported` theme support arg to specify the default `supported_templates` option (make it `false` if the theme support arg is not `'all'`, and if it is an array use it to merge with the default `supported_templates` value.

## Testing

To test, it's helpful to write out this JSON out to a file (e.g. `options.json`):

```json
{"theme_support":"transitional","supported_post_types":["post"],"analytics":[],"all_templates_supported":false,"supported_templates":["is_singular"],"version":"1.5.5"}
```

And then load up the old plugin options via:

```bash
cat options.json | wp option set amp-options --format=json
```

It's also helpful to have these plugins to activate/deactivate to see how they impact the migration:

```php
<?php
/**
 * Plugin Name: AMP Add Post Type Support for Pages
 */

add_action( 'init', function() {
	add_post_type_support( 'page', 'amp' );
} );
```

```php
<?php
/**
 * Plugin Name: AMP Add Theme Support Templates Supported All
 */

add_action(
	'after_setup_theme',
	function () {
		add_theme_support( 'amp', [ 'templates_supported' => 'all' ] );
	}
);
```

```php
<?php
/**
 * Plugin Name: AMP Add Theme Support Templates Supported Selectively
 */

add_action(
	'after_setup_theme',
	function () {
		add_theme_support(
			'amp',
			[
				'templates_supported' => [
					'is_search' => true,
					'is_404'    => false,
					'is_date'   => true,
				],
			]
		);
	}
);
```

```php
<?php
/**
 * Plugin Name: AMP Skip Post All
 */

add_filter( 'amp_skip_post', '__return_true' );
```

## Other

* Remove obsolete code for password-protected posts which was missed when resolving #3428.

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

westonruter committed:
Default to paired when amp theme support added

westonruter committed:
Eliminate use of theme support flags and amp_supportable_templates filter to force templates to be supported

westonruter committed:
Eliminate use of post type support to control post type support

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into update/theme-support-paired-default

* 'develop' of github.com:ampproject/amp-wp: (94 commits)
  Improve formatting of active theme notice
  Open ecosystem page in a new window
  Remove dead code and update translator comment formatting
  Settings page CSS updates and test fixes
  Restore supported templates fieldset without obsolete wrapper
  Remove unused variable
  Add escaping to menu strings
  Make details prop a string instead of a node
  Escaping and minor updates to OptionsMenu.php
  Remove redundant isset check
  Rename reader-mode-override.js
  Comment formatting fix
  Change registration action for GoogleFonts class
  Restore ElementName
  Change ElementName to elementName and lint fixes
  Remove unnecessary default option provided via filter elsewhere
  Supported templates visibility updates
  Add todo indicating Plugin Suppression will be handled over REST in the future
  Increase width of status column in plugin suppression table
  Improve readability around mobile redirection visibility
  ...

westonruter committed:
Remove test kept after merge conflict

westonruter committed:
Fix tests for changes to post type support

westonruter committed:
Add tests for migrating supported_post_types

westonruter committed:
Migrated supported_templates to be stored as associative array

westonruter committed:
Remove redundant isset checks with alternative to appease PhpStan

westonruter committed:
Remove _doing_it_wrong() since templates_supported is supported only for default value

westonruter committed:
Add missing semicolons from JS

westonruter commented on includes/options/class-amp-options-manager.php:
```diff
@@ -155,6 +175,16 @@ public static function get_options() {
 			$options[ Option::THEME_SUPPORT ] = $defaults[ Option::THEME_SUPPORT ];
 		}
 
+		// Migrate supported post types.
+		if ( isset( $options[ Option::SUPPORTED_POST_TYPES ][0] ) ) {
+			$options[ Option::SUPPORTED_POST_TYPES ] = array_fill_keys( $options[ Option::SUPPORTED_POST_TYPES ], true );
+		}
+
+		// Migrate supported templates.
+		if ( isset( $options[ Option::SUPPORTED_TEMPLATES ][0] ) ) {
+			$options[ Option::SUPPORTED_TEMPLATES ] = array_fill_keys( $options[ Option::SUPPORTED_TEMPLATES ], true );
+		}
```
I think this actually needs to take into account the `version`, and make sure the theme args are accounted for here as well.

westonruter committed:
Make sure toggling all themes supported does not destroy individual supported template preferences

westonruter committed:
Migrate theme support flags to options when upgrading

westonruter committed:
Defer doing amp_plugin_update until end of after_setup_theme

westonruter commented on includes/options/class-amp-options-manager.php:
```diff
@@ -27,10 +27,10 @@ class AMP_Options_Manager {
 		Option::THEME_SUPPORT           => AMP_Theme_Support::READER_MODE_SLUG,
-		Option::SUPPORTED_POST_TYPES    => [ 'post' ],
+		Option::SUPPORTED_POST_TYPES    => [ 'post' => true ],
 		Option::ANALYTICS               => [],
 		Option::ALL_TEMPLATES_SUPPORTED => true,
-		Option::SUPPORTED_TEMPLATES     => [ 'is_singular' ],
+		Option::SUPPORTED_TEMPLATES     => [ 'is_singular' => true ],
```
I'm second-guessing why using an associative array is necessary. If an item is absent, then this is also treated as being unsupported. So a simple array of strings should be fine.

pierlon replied:
Agreed.

westonruter replied:
Undone in 3a326c3

pierlon commented on includes/amp-helper-functions.php:
```diff
@@ -143,14 +141,22 @@ function () {
+		// This waits to happen until after_setup_theme to ensure that amp theme support and amp post type support have all been added.
+		add_action(
+			'after_setup_theme',
+			function () use ( $old_version ) {
```
```suggestion
			static function () use ( $old_version ) {
```

westonruter committed:
Remove obsolete checks for password-protected posts

See #3428

westonruter committed:
Remove obsolete error codes for page-on-front and page-for-posts

westonruter committed:
Fix tests and finish changes for get_status_and_errors

pierlon commented on includes/admin/class-amp-post-meta-box.php:
```diff
@@ -338,29 +340,20 @@ public static function get_status_and_errors( $post ) {
 	 * @return array $error_messages The error messages, as an array of strings.
 	 */
-	public function get_error_messages( $status, $errors ) {
+	public function get_error_messages( $errors ) {
```
Do you think `AMP_Post_Meta_Box::get_status_and_errors` should be refactored into 2 separate methods? The returned status is now only being used in 1 instance (`AMP_Post_Meta_Box::get_amp_enabled_rest_field`), and the set of errors are used twice elsewhere (`AMP_Post_Meta_Box::enqueue_block_assets` and `AMP_Post_Meta_Box::render_status`).

westonruter replied:
It's being used both in `AMP_Post_Meta_Box::sanitize_status()` and `AMP_Post_Meta_Box::render_status()`, in the latter case it _appears_ to not be used only because the usage is in the `include` file at the end of the method.

Nevertheless, I'm not in love with `get_status_and_errors`.

pierlon commented on src/Admin/OptionsMenu.php:
```diff
@@ -440,27 +430,16 @@ private function list_template_conditional_options( $options, $parent = null ) {
+						id="<?php echo esc_attr( $element_id ); ?>"
+						<?php checked( $supported ); ?>
+					>
+					<input type="hidden" name="<?php echo esc_attr( $element_name ); ?>" value="<?php echo esc_attr( wp_json_encode( ! empty( $supported_templates[ $id ] ) ) ); ?>">
```
Why handle the checkbox value through a hidden input?

westonruter replied:
It's because a checkbox that is not checked will not send any post data. So I modified the JS to toggle the hidden input with a boolean value. This is intended to be temporary since @johnwatkins0 will make it fully Reactive in #5015 .

westonruter replied:
In other words, if no checkboxes are checked, then there will be no input vars to indicate the user is intending to make a change for these. This will not be an issue once moving to the REST API. I suppose my change here is overkill, however.

westonruter replied:
Reverted/revised in 3a326c3b191d304022dda30d669c107f3da79fcc.

pierlon commented on tests/php/test-amp-helper-functions.php:
```diff
@@ -86,6 +87,101 @@ public function return_example_url( $url, $post_id ) {
+		$this->assertEquals( 0, did_action( 'amp_init' ) );
+		amp_init();
+		$this->assertEquals( 1, did_action( 'amp_init' ) );
+		$this->assertEquals( 10, has_filter( 'allowed_redirect_hosts', [ 'AMP_HTTP', 'filter_allowed_redirect_hosts' ] ) );
```
Why check for this specific filter?

westonruter replied:
It's just something that is added in that method, so just confirming that the filter is indeed added as expected.

pierlon commented on tests/php/test-amp-helper-functions.php:
```diff
@@ -86,6 +87,101 @@ public function return_example_url( $url, $post_id ) {
+
+		add_action(
+			'after_setup_theme',
+			function () {
```
```suggestion
			static function () {
```

pierlon commented on tests/php/test-amp.php:
```diff
@@ -69,6 +69,11 @@ public function get_amp_is_canonical_test_data() {
 				true,
 				AMP_Theme_Support::STANDARD_MODE_SLUG,
 			],
+			'no_paired_arg'               => [
```
Seems to be a duplicate of the `no_args` above.

westonruter replied:
Removed in a39c21237.

pierlon commented on tests/php/test-class-amp-options-manager.php:
```diff
@@ -268,6 +298,174 @@ public function test_get_and_set_options() {
+	 * @covers AMP_Options_Manager::get_options()
+	 */
+	public function test_get_options_migration_supported_post_types_defaults() {
+		foreach ( get_post_types() as $post_type ) {
```
This already happens during `tearDown`, no?

westonruter replied:
Yes, but I intended to make sure that no other test added the post type support immediately before.

pierlon commented on tests/php/test-class-amp-options-manager.php:
```diff
@@ -268,6 +298,174 @@ public function test_get_and_set_options() {
+	 * @covers AMP_Options_Manager::get_options()
+	 */
+	public function test_get_options_migration_supported_templates() {
+		// Make sure that the old simple list of post types gets migrated to an associative array.
```
```suggestion
		// Make sure that the old simple list of supported templates gets migrated to an associative array.
```

westonruter replied:
Made obsolete in 8c51a9dcf

westonruter committed:
Revamp AMP Unavailable notice in post editor

* Hide if user can't manage_options anyway.
* Open links to setting screen in new window.
* Add "AMP Unavailable" heading to notice.
* Use <details> to contain error messages in notice in block editor.
* Use notice as opposed to warning.

westonruter committed:
Restore using non-associative arrays to store supported post types and supported templates

westonruter committed:
Remove obsolete assertion

westonruter committed:
Use static closures

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Remove duplicate test data

westonruter committed:
Remove invalid todo

westonruter commented on includes/admin/class-amp-post-meta-box.php:
```diff
@@ -338,40 +344,33 @@ public static function get_status_and_errors( $post ) {
+		$settings_screen_url = admin_url( 'admin.php?page=' . AMP_Options_Manager::OPTION_NAME );
+
 		$error_messages = [];
-		if ( in_array( 'status_immutable', $errors, true ) ) {
-			if ( self::ENABLED_STATUS === $status ) {
-				$error_messages[] = __( 'Your site does not allow AMP to be disabled.', 'amp' );
-			} else {
-				$error_messages[] = __( 'Your site does not allow AMP to be enabled.', 'amp' );
-			}
-		}
```
Byebye!

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into update/theme-support-paired-default

* 'develop' of github.com:ampproject/amp-wp:
  Update dependency dealerdirect/phpcodesniffer-composer-installer to v0.7.0
  Harden ReaderThemesTest when custom theme directories present
  Update dependency sirbrillig/phpcs-variable-analysis to v2.8.3
  Skip test in WP 4.9
  Fix grammar typo
  Remove escaping of title in legacy post templates (#5026)
  Skip test when Twenty Twenty is not installed
  Prevent Cover Template section in AMP Customizer from appearing when Twenty Twenty used as Reader theme
  Add is_theme_overridden method on ReaderThemeLoader service
  Improve detection for whether or not ReaderThemeSwitcher is enabled
  Make ReaderThemeLoaer unconditional and add method to obtain originally-active theme
  Prevent sanitization of common style elements in the Customizer preview
  Fix lint issue in test
  Fix lint issue
  Add test
  Prevent sanitization of styles in some Core themes requried for some Customizer features to work correctly

pierlon commented on includes/amp-helper-functions.php:
```diff
@@ -143,14 +141,22 @@ function () {
-		AMP_Options_Manager::update_option( Option::VERSION, AMP__VERSION );
+		// This waits to happen until after_setup_theme to ensure that amp theme support and amp post type support have all been added.
+		add_action(
+			'after_setup_theme',
```
The `after_setup_theme` action would have already been fired, no? If I recall correctly, the `init` action comes after `after_setup_theme`.

westonruter replied:
Humm, yeah, you're right. I need to re-check this.

westonruter replied:
Fixed in 7df6a05378e8b7e6fdeefac9a159a6e7117257d3

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into update/theme-support-paired-default

* 'develop' of github.com:ampproject/amp-wp: (36 commits)
  Fix typo in comment
  Update dependency @wordpress/plugins to v2.20.1
  Improve copy for panel description and control notice
  Handle case where user resets devtools option back to original
  Remove extraneous em unit from line-height
  Add link to customize non-AMP version
  Update AMP panel description in Customizer
  Update dependency eslint-plugin-react-hooks to v4.0.8
  Add AMP panel description to the AMP Customizer root
  Update dependency @wordpress/scripts to v12.1.1
  Update dependency @wordpress/components to v10.0.1
  Update dependency eslint-plugin-react to v7.20.3
  Update dependency eslint-plugin-jest to v23.18.0
  Update dependency eslint-plugin-jsdoc to v29.2.0
  Add notification to the menus panel explaining menus are shared
  Use wp.customize alias
  Simplify logic for removing Cover Template section
  Remove the Homepage Settings section in AMP Customizer for a Reader theme
  Add notice to all Customizer controls for option settings
  Re-organize functions in amp-customize-controls
  ...

westonruter committed:
Use end of init instead of end of after_setup_theme to do DB upgrade

pierlon commented on assets/css/src/amp-block-editor.css:
```diff
@@ -33,3 +33,16 @@
 }
+
+.amp-unavailable-notice {
+	width: 100%;
```
```suggestion
	width: 100%;
	margin: 5px 0 2px;
```

pierlon replied:
Before | After
---|---
![image](https://user-images.githubusercontent.com/16200219/87372881-22464780-c578-11ea-9359-b07b7cdb6839.png) |  ![image](https://user-images.githubusercontent.com/16200219/87372828-03e04c00-c578-11ea-981e-cc2011826a5a.png)

pierlon commented on assets/css/src/amp-block-editor.css:
```diff
@@ -33,3 +33,16 @@
+
+.amp-unavailable-notice details > summary {
+	font-weight: bold;
+	user-select: none;
```
```suggestion
	user-select: none;
	cursor: pointer;
```

pierlon reviewed (commented):
Functionality works as intended, only a minor visual fix I'd like to suggest.

westonruter committed:
Improve styling of AMP Unavailable notice

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter commented:
Just noticed that WordPress 5.5 removes the blue background for the notices. The margins are also a bit off, but this is out of scope of this PR. For example, the Featured Image notice:

![image](https://user-images.githubusercontent.com/134745/87384081-bc45ca00-c54f-11ea-9b32-d00b39b2540a.png)

westonruter commented:
Two follow-up items I've addressed in #5042:

* `AMP_Core_Theme_Sanitizer::extend_theme_support()` would fail to add `amp` theme support if a core theme doesn't have any special theme support args. This resulted in Twenty Sixteen getting Transitional mode by default, but Twenty Twenty did not.
* When a child theme of a core theme is active, we should _not_ default to Transitional or Standard because all bets are off as to whether it actually supports AMP. Fixed in f8c8390. We'll only be able to change that once we have the theme scan feature (#4795).
